### PR TITLE
`sort`: Partially migrate from THC to ATen, replace the thrust path with cub

### PR DESCRIPTION
**Pull request by zasdfgbnm**
The thrust path of `torch.sort` in THC is rewritten and replaced with cub in ATen. The original algorithm is followed, but since cub does not offer custom compare operator, I have to change it a bit to 2 sort + gather.

Note: tensor larger than 2^31 elements is supported, but the dimension being sorted can not go beyond 2^31.

Related: https://github.com/pytorch/pytorch/pull/50887 https://github.com/pytorch/pytorch/issues/24637

Benchmark:

```python
import torch
import itertools

for i in range(1000):
    torch.arange(100000, device='cuda')

def run50_sync(f):
    for _ in range(50):
        f()
    torch.cuda.synchronize()

for i, j in itertools.product([512, 4096, 8192], repeat=2):
    print(i,j)
    t = torch.randn(i, j, device='cuda')
    torch.cuda.synchronize()
    %timeit run50_sync(lambda: torch.sort(t))
    torch.cuda.synchronize()
    %timeit run50_sync(lambda: torch.sort(t, dim=0))
    print()
```

Before
```
512 512
3.91 ms ± 8.53 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
4.87 ms ± 5.06 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

512 4096
70.5 ms ± 29.4 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
32.7 ms ± 14.2 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

512 8192
142 ms ± 21.4 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
64.4 ms ± 94.9 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

4096 512
26.8 ms ± 1.68 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
82.2 ms ± 13.4 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

4096 4096
606 ms ± 178 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
722 ms ± 94.8 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

4096 8192
1.28 s ± 157 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.54 s ± 500 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

8192 512
53.5 ms ± 73.1 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
168 ms ± 39.4 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

8192 4096
1.28 s ± 236 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.54 s ± 272 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

8192 8192
2.69 s ± 741 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
3.28 s ± 549 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

After
```
512 512
4.02 ms ± 28.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
5 ms ± 15.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

512 4096
40.7 ms ± 74.2 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
33.9 ms ± 186 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

512 8192
71.7 ms ± 636 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
66.4 ms ± 163 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

4096 512
27.6 ms ± 27.8 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
46.6 ms ± 101 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

4096 4096
262 ms ± 1.14 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
321 ms ± 1.32 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

4096 8192
520 ms ± 5.47 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
661 ms ± 853 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

8192 512
54.6 ms ± 133 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
83.2 ms ± 320 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

8192 4096
521 ms ± 1.06 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
645 ms ± 1.47 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

8192 8192
1.04 s ± 2.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.34 s ± 541 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
Haven't tested the `test_sort_large` yet, need a larger GPU

**Comment by ezyang** [comment]
importing for better diff view, please commandeer if you plan to land

**Comment by ezyang** [comment]
ACKing native_Functions.yaml changes

File: aten/src/ATen/native/cuda/Sort.cu [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+  AT_DISPATCH_ALL_TYPES_AND(kHalf, self_.scalar_type(), "sort", [&]{
+    const scalar_t *self_ptr = self_.data_ptr<scalar_t>();
+    scalar_t *tmp_ptr = tmp.data_ptr<scalar_t>();
+    auto values_ptr = reinterpret_cast<scalar_t *>(values_ptr_);
+    int64_t remaining = numel;
+    while (remaining > 0) {
+      int64_t n = std::min(remaining, nbatch);
+      int64_t segment_bits = std::max<int64_t>(1L, static_cast<int64_t>(std::ceil(std::log2(n / nsort))));
+      at::cuda::cub::sort_pairs(
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Sort.cu:182]: I suspect instead of doing 2 sorts with the same key, it would be faster to do a single sort with arange as value, and then gather indices and segment ids according to permuted arange? thrust::gather hopefully should have async apis and not create device cache, if not, writing a simple gather kernel should not be awful?

Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Sort.cu]: Yes, the two sort algorithm is faster. Now the new code is unconditionally faster than the old one. And this PR should be ready.

Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Sort.cu]: New benchmark updated in https://github.com/pytorch/pytorch/pull/54626#issue-600072099

File: aten/src/ATen/cuda/cub.cuh [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+  auto allocator = c10::cuda::CUDACachingAllocator::get();
+  c10::DataPtr keys_out_owner;
+  c10::DataPtr values_out_owner;
+
+  if (keys_out == nullptr) {
+    keys_out_owner = allocator->allocate(n * sizeof(key_t));
+    keys_out = reinterpret_cast<key_t *>(keys_out_owner.get());
+  }
+  if (keys_out == nullptr) {
```
Comment (ngimel) [inline aten/src/ATen/cuda/cub.cuh:70]: should it be values_out == nullptr?

Comment (zasdfgbnm) [reply aten/src/ATen/cuda/cub.cuh]: Good catch! fixed

File: aten/src/ATen/native/cuda/Sort.cu [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+  AT_DISPATCH_ALL_TYPES_AND(kHalf, self_.scalar_type(), "sort", [&]{
+    const scalar_t *self_ptr = self_.data_ptr<scalar_t>();
+    auto values_ptr = reinterpret_cast<scalar_t *>(values_ptr_);
+    int64_t remaining = numel;
+    while (remaining > 0) {
+      int64_t n = std::min(remaining, nbatch);
+      int64_t nsegments = n / nsort;
+      int64_t segment_bits = std::max<int64_t>(1L, static_cast<int64_t>(std::ceil(std::log2(nsegments))));
+
+      auto indices_and_segment = at::empty({nsegments, nsort, 2}, indices.options());
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Sort.cu:166]: why do you need a long here? values are guaranteed to fit in `int`.

Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Sort.cu]: Yes! And taking advantage that, I can get another 1.8x speedup!

Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Sort.cu]: New benchmark updated in https://github.com/pytorch/pytorch/pull/54626#issue-600072099

File: test/test_sort_and_select.py [hunk 4]
```diff
@@ -48,70 +52,71 @@ def check_order(a, b):
             self.assertEqual(len(seen), size)
 
     def test_sort(self, device):
```
Comment (zasdfgbnm) [inline test/test_sort_and_select.py:54]: "Hide whitespace changes" on GitHub will provide better diff view for review

File: test/test_sort_and_select.py [hunk 5]
```diff
@@ -135,7 +147,60 @@ def test_stable_sort(self, device, dtype):
                 torch.arange(start=1, end=2 * ncopies, step=2, device=device)
             )
 
+    @onlyCUDA
+    @dtypes(torch.uint8)
+    @largeTensorTest('200GB')  # Unfortunately 80GB A100 is not large enough
+    def test_sort_large(self, device, dtype):
```
Comment (ngimel) [inline test/test_sort_and_select.py:153]: so basically we can't test a tensor with more than 2**31 elements? lol

**Review by ngimel — APPROVED** [review]
